File: 892-shortest-subarray-with-sum-at-least-k/shortest-subarray-with-sum-at-least-k.cs


public class Solution {
    public int ShortestSubarray(int[] nums, int k) {
        int n = nums.Length;
        var pq = new SortedSet<(long Sum, int Index)>();
        long sum = 0;
        int minlen = n + 1;

        for (int i = 0; i < n; i++) {
            sum += nums[i];
            if (sum >= k) {
                minlen = Math.Min(minlen, i + 1);
            }
            while (pq.Count > 0 && sum - pq.Min.Sum >= k) {
                minlen = Math.Min(minlen, i - pq.Min.Index);
                pq.Remove(pq.Min);
            }
            pq.Add((sum, i));
        }

        return minlen > n ? -1 : minlen;
    }
}

File: 892-shortest-subarray-with-sum-at-least-k/shortest-subarray-with-sum-at-least-k.cpp
class Solution {
public:
    int shortestSubarray(vector<int>& nums, int k) {
        int n = nums.size();

        priority_queue<pair<long long, int>, vector<pair<long long, int>>, greater<pair<long long, int>>> pq;

        long long sum = 0;
        int minlen = n + 1;

        for(int i =0; i<n; i++){
            sum += nums[i];

            if(sum >= k){
                minlen = min(minlen, i + 1);
            }

            while(!pq.empty() && sum - pq.top().first >= k){
                minlen = min(minlen, i - pq.top().second ); pq.pop();
            }

            pq.push({sum, i});
        }

        return minlen > n ? -1 : minlen;

    }
};